subprojects/core/src/main/java/org/gradle/internal/classpath/InstrumentingTransformer.java

```diff
@@ -133,136 +148,159 @@ public void visitEnd() {
+                        _INVOKESTATIC(INSTRUMENTED_TYPE, "getLong", RETURN_LONG_FROM_STRING_PRIMITIVE_LONG_STRING);
                         return;
                     }
                     if (descriptor.equals(RETURN_LONG_FROM_STRING_LONG)) {
-                        visitLdcInsn(Type.getObjectType(className).getClassName());
-                        super.visitMethodInsn(Opcodes.INVOKESTATIC, INSTRUMENTED_TYPE.getInternalName(), "getLong", RETURN_LONG_FROM_STRING_LONG_STRING, false);
+                        _LDC(binaryClassNameOf(className));
+                        _INVOKESTATIC(INSTRUMENTED_TYPE, "getLong", RETURN_LONG_FROM_STRING_LONG_STRING);
                         return;
                     }
                 } else if (owner.equals(BOOLEAN_TYPE.getInternalName()) && name.equals("getBoolean") && descriptor.equals(RETURN_PRIMITIVE_BOOLEAN_FROM_STRING)) {
-                    visitLdcInsn(Type.getObjectType(className).getClassName());
-                    super.visitMethodInsn(Opcodes.INVOKESTATIC, INSTRUMENTED_TYPE.getInternalName(), "getBoolean", RETURN_PRIMITIVE_BOOLEAN_FROM_STRING_STRING, false);
+                    _LDC(binaryClassNameOf(className));
+                    _INVOKESTATIC(INSTRUMENTED_TYPE, "getBoolean", RETURN_PRIMITIVE_BOOLEAN_FROM_STRING_STRING);
                     return;
                 } else if (owner.equals(className) && name.equals(CREATE_CALL_SITE_ARRAY_METHOD) && descriptor.equals(RETURN_CALL_SITE_ARRAY)) {
-                    super.visitMethodInsn(Opcodes.INVOKESTATIC, className, INSTRUMENTED_CALL_SITE_METHOD, RETURN_CALL_SITE_ARRAY, false);
+                    _INVOKESTATIC(className, INSTRUMENTED_CALL_SITE_METHOD, RETURN_CALL_SITE_ARRAY);
                     return;
                 }
             }
             super.visitMethodInsn(opcode, owner, name, descriptor, isInterface);
         }
 
+        private String binaryClassNameOf(String className) {
+            return getObjectType(className).getClassName();
+        }
+
         @Override
         public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootstrapMethodHandle, Object... bootstrapMethodArguments) {
-            if (descriptor.endsWith(")" + Type.getType(Action.class).getDescriptor()) && bootstrapMethodHandle.getOwner().equals(Type.getType(LambdaMetafactory.class).getInternalName()) && bootstrapMethodHandle.getName().equals("metafactory")) {
-                Handle altMethod = new Handle(Opcodes.H_INVOKESTATIC,
-                    Type.getType(LambdaMetafactory.class).getInternalName(),
+            if (isGradleLambdaDescriptor(descriptor) && bootstrapMethodHandle.getOwner().equals(getType(LambdaMetafactory.class).getInternalName()) && bootstrapMethodHandle.getName().equals("metafactory")) {
+                Handle altMethod = new Handle(
+                    H_INVOKESTATIC,
+                    getType(LambdaMetafactory.class).getInternalName(),
```

Comment:
we should probably turn this and the method descriptor below into constants, similar to what happens for other descriptors and type names (I should have done that initially but forgot).

Reply:
Done.